static/js/buddy_data.js

```diff
@@ -247,30 +247,79 @@ function maybe_shrink_list(user_ids, filter_text) {
-        var person = people.get_person_from_user_id(user_id);
-
-        if (person) {
-            // if the user is bot, do not show in presence data.
-            if (person.is_bot) {
-                return false;
+    // If or condition evaluates to force, follow normal "all user" path
+    if (use_member_list && (use_pm_with_filter || use_stream_filter)) {
+        if (use_pm_with_filter) {
+            var user_emails = filter.operands('pm-with')[0].split(',');
+            user_ids = _.reduce(user_emails, function (memo, user_email) {
+                var user_id = people.get_user_id(user_email);
+                var person = people.get_person_from_user_id(user_id);
+                // if the user is bot, do not show in presence data.
+                if (person && !person.is_bot) {
+                    memo.push(user_id);
+                }
+                return memo;
+            }, []);
+            // Add current user to buddy list
+            var me = people.my_current_user_id();
+            if (!_.contains(user_ids, me)) {
+                user_ids.push(me);
             }
+        } else { // use_stream_filter must be true
+            user_ids = filter_user_ids(filter_text, people.get_active_user_ids());
+            var sub = filter.operands('stream')[0];
+            user_ids = _.filter(user_ids, function (user_id) {
+                var person = people.get_person_from_user_id(user_id);
+                if (person) {
+                    // if the user is bot, do not show in presence data.
+                    if (person.is_bot) {
+                        return false;
+                    }
+                }
+                if (sub && !stream_data.is_user_subscribed(sub, person.user_id)) {
+                    return false;
+                }
+                return true;
+            });
+        }
+        if (filter_text) {
+            user_ids = filter_user_ids(filter_text, user_ids);
+        }
+    } else {
+        if (filter_text) {
+            // If there's a filter, and we're not using the members list
+            // then select from all users, not just those recently active.
+            user_ids = filter_user_ids(filter_text, people.get_active_user_ids());
+        } else {
+            // From large realms, the user_ids in presence may exclude
+            // users who have been idle more than three weeks.  When the
+            // filter text is blank, we show only those recently active users.
+            user_ids = presence.get_user_ids();
         }
-        return true;
-    });
 
+        user_ids = _.filter(user_ids, function (user_id) {
+            var person = people.get_person_from_user_id(user_id);
+            if (person) {
+                // if the user is bot, do not show in presence data.
+                if (person.is_bot) {
+                    return false;
+                }
+            }
+            return true;
+        });
+    }
```

Comment:
This function, in general, feels really hard to read and understand.  I think there's a few reasonable options:
* Probably best, take some of these procedures that need to be used in multiple code paths, like filtering out bots, and turn them into functions in a prep commit, so that the end result is readable.
* Use the early return pattern to have the pre-existing code look the same as before to make the diff understandable
* Just move to two functions for the two different use cases; they are completely different algorithms